### PR TITLE
Adds MMLU benchmark support

### PR DESCRIPTION
**Pull request by patemotter**
Adds mmlu as a supported dataset. 

Tested with local runs.

### CODE REVIEW

File: scripts/agent/benchmark_dataset.py [hunk 1]
```diff
@@ -10,13 +10,15 @@
   - BurstGPT
   - HuggingFace
   - VisionArena
+  - MMLU
```
Comment (QiliangCui) [inline scripts/agent/benchmark_dataset.py:13]: Hm.. this file is a temp solution for customized dataset. If MMLU is something we want to officially support, can you add it to https://github.com/vllm-project/vllm/blob/main/benchmarks/benchmark_dataset.py? 

This file is a fork of https://github.com/vllm-project/vllm/blob/main/benchmarks/benchmark_dataset.py and I made some changes for an urgent experiment when I didn't have time to convince the vllm community.

File: scripts/agent/docker_run_bm.sh [hunk 2]
```diff
@@ -78,13 +78,20 @@ docker run \
  $image_tag tail -f /dev/null
 
 # =============== temp solution ===============
-if [ "$DATASET" = "custom-token" ]; then
-  echo "use temp solution to hack custom-token dataset"
+if [ "$DATASET" = "custom-token" ] || [ "$DATASET" = "mmlu" ]; then
```
Comment (QiliangCui) [inline scripts/agent/docker_run_bm.sh:81]: here is the thing. I think we should add that dataset to the official support unless they strongly push back. 

The solution here was for an urgent test.

File: scripts/agent/benchmark_serving.py [hunk 3]
```diff
@@ -72,6 +72,7 @@
     ShareGPTDataset,
     SonnetDataset,
     VisionArenaDataset,
+    MMLUDataset,
```
Comment (QiliangCui) [inline scripts/agent/benchmark_serving.py:75]: Let's try to push to https://github.com/vllm-project/vllm/blob/main/benchmarks/benchmark_serving.py unless they strongly disagree. 

I previously did this just for an urgent experiment.

File: scripts/agent/benchmark_serving.py [hunk 4]
```diff
@@ -753,6 +769,17 @@ def main(args: argparse.Namespace):
             tokenizer=tokenizer, num_requests=args.num_prompts
         )
 
+    elif args.dataset_name == "mmlu":
+        dataset = MMLUDataset(
+            dataset_path=args.dataset_path,
```
Comment (QiliangCui) [inline scripts/agent/benchmark_serving.py:774]: where is the dataset_path set?

File: scripts/agent/benchmark_dataset.py (hunk 1 above)
Comment (mailvijayasingh) [reply scripts/agent/benchmark_dataset.py:13]: yes, we want to support this and while we will work on supporting this on vllm parallely, I think we should merge this as a temp solution. 

We can later change it to a generic custom dataset path in place of `MMLU`. How does that sound to you?

Comment (QiliangCui) [reply scripts/agent/benchmark_dataset.py:13]: Let's create a TODO bug?

File: cases/hourly_jax.csv [hunk 5]
```diff
@@ -4,3 +4,4 @@ v6e-1,Qwen/Qwen2.5-7B-Instruct,128,512,1,2048,sonnet,1800,128
 v6e-8,Qwen/Qwen2.5-14B-Instruct,128,512,8,2048,sonnet,1800,128
 v6e-8,Qwen/Qwen2.5-32B,128,512,8,2048,sonnet,1800,128
 v6e-1,meta-llama/Llama-3.1-8B-Instruct,256,1024,1,2048,sonnet,1800,128
+v6e-8,meta-llama/Llama-3.3-70B-Instruct,128,1024,1,2048,mmlu,1800,128
```
Comment (QiliangCui) [inline cases/hourly_jax.csv:7]: when this is merged, the scheduler will pick up this immediately and start to run test on this before the agent machine is refreshed. So, you will see errors. 

I will do an agent deployment then. if things works well, the error should go away then.

File: scripts/agent/docker_run_bm.sh [hunk 6]
```diff
@@ -131,7 +139,7 @@ gsutil cp $LOG_ROOT/* $REMOTE_LOG_ROOT
 # 
 if [[ -z "$throughput" || ! "$throughput" =~ ^[0-9]+([.][0-9]+)?$ ]]; then
   echo "Failed to get the throughput"
-  exit 0
+  exit 1
```
Comment (QiliangCui) [inline scripts/agent/docker_run_bm.sh:142]: please return 0. 

the upper code will handle the error by checking if result exists.

File: scripts/agent/docker_run_bm.sh [hunk 7]
```diff
@@ -172,4 +180,5 @@ P99TTFT=$P99TTFT
 P99ETEL=$P99ETEL
 OutputTokenThroughput=$output_token_throughput
 TotalTokenThroughput=$total_token_throughput
+AccuracyMetrics='$AccuracyMetricsJSON'
```
Comment (QiliangCui) [inline scripts/agent/docker_run_bm.sh:183]: I see you updated the spanner. 

Please check in the updating script as script in https://github.com/QiliangCui/bm-infra/tree/main/database

File: scripts/agent/run_bm.sh [hunk 8]
```diff
@@ -32,15 +32,14 @@ mkdir "$WORKSPACE/log"
...
-#
-# create sonnet_4x
-#
-echo "Create sonnet_4x.txt"
-echo "" > benchmarks/sonnet_4x.txt
-for _ in {1..4}
- do
-  cat benchmarks/sonnet.txt >> benchmarks/sonnet_4x.txt
-done
+if [ "$DATASET" = "sonnet" ]; then
```
Comment (QiliangCui) [inline scripts/agent/run_bm.sh:35]: did you verify this with "sonnet"?

File: scripts/agent/run_bm.sh [hunk 9]
```diff
@@ -141,77 +152,78 @@ run_benchmark(){
 
   throughput=$(grep "Request throughput (req/s):" "$BM_LOG" | sed 's/[^0-9.]//g')
   p99_e2el=$(grep "P99 E2EL (ms):" "$BM_LOG" | awk '{print $NF}')
-
   echo "throughput: $throughput, P99 E2EL:$p99_e2el"
   echo
-
-  # return
-  echo "$throughput $p99_e2el"
+
```
Comment (QiliangCui) [inline scripts/agent/run_bm.sh:157]: why delete this?

I think this will break the following steps.

the line `throughput p99_e2el < <(run_benchmark "inf" | tail -n 1) ` rely on this echo.

File: scripts/agent/run_bm.sh [hunk 10]
```diff
@@ -141,77 +152,78 @@ run_benchmark(){
...
-
   echo "throughput: $throughput, P99 E2EL:$p99_e2el"
   echo
-
-  # return
-  echo "$throughput $p99_e2el"
+  
 }
-read throughput p99_e2el < <(run_benchmark "inf" | tail -n 1) 
+if [ "$DATASET" = "sonnet" ]; then
```
Comment (QiliangCui) [inline scripts/agent/run_bm.sh:159]: why? 

the logic below is using to find best result when there is e2e latency requirement. it is not just for sonnet.

File: scripts/agent/run_bm.sh (hunk 9 above)
Comment (patemotter) [reply scripts/agent/run_bm.sh:157]: So then one of these is redundant but it't ambiguous about which one is actually needed? Or is there a reason we print out the throughput twice in 5 lines?

File: scripts/agent/run_bm.sh (hunk 10 above)
Comment (patemotter) [reply scripts/agent/run_bm.sh:159]: It is currently just for sonnet though since that was the only dataset supported. And this was being executed for mmlu each time because it makes an assumption about how big the e2el should be even if that isn't already known.

File: scripts/agent/benchmark_serving.py (hunk 4 above)
Comment (patemotter) [reply scripts/agent/benchmark_serving.py:774]: It's defined in `run_bm.sh` and the data itself is being downloaded from GCS in `docker_run_bm.sh`. 

```
  elif [ "$DATASET" = "mmlu" ]; then
    python benchmarks/benchmark_serving.py \
      --backend vllm \
      --model $MODEL \
      --request-rate $request_rate \
      --dataset-name mmlu \
      --dataset-path /workspace/dataset \
      --mmlu-num-shots 5 \
      --mmlu-method HELM \
      --num-prompts ${NUM_PROMPTS} \
      --percentile-metrics ttft,tpot,itl,e2el \
      --ignore-eos > "$BM_LOG" 2>&1
```

File: scripts/agent/docker_run_bm.sh (hunk 7 above)
Comment (patemotter) [reply scripts/agent/docker_run_bm.sh:183]: Added.

File: scripts/agent/run_bm.sh (hunk 8 above)
Comment (patemotter) [reply scripts/agent/run_bm.sh:35]: I did, but I am re-running with the recent changes.

Comment (patemotter) [reply scripts/agent/run_bm.sh:35]: Confirmed working with latest changes.

File: scripts/agent/run_bm.sh (hunk 9 above)
Comment (patemotter) [reply scripts/agent/run_bm.sh:158]: I have added it back in for the time being.

File: scripts/agent/benchmark_serving.py [hunk 11]
```diff
@@ -221,10 +223,16 @@ def calculate_metrics(
...
     ttfts: list[float] = []
     e2els: list[float] = []
+    correct_predictions = 0
     for i in range(len(outputs)):
         if outputs[i].success:
             output_len = outputs[i].output_tokens
 
+            # Accuracy check
+            if input_requests[i].completion is not None:
+                if input_requests[i].completion in outputs[i].generated_text:
```
Comment (QiliangCui) [inline scripts/agent/benchmark_serving.py:233]: Is this a standard way to calculate accuracy? 

I highly recommend we put this logic to vllm main to have a broader review for change like this.

File: scripts/agent/docker_run_bm.sh [hunk 12]
```diff
@@ -88,13 +88,20 @@ docker run \
...
+  if [ "$DATASET" = "custom-token" ]; then
+    # Download flat files for custom-token
+    gsutil -m cp gs://$GCS_BUCKET/dataset/*.* ./artifacts/dataset/
+  elif [ "$DATASET" = "mmlu" ]; then
+    # Download mmlu directory recursively
+    gsutil -m cp -r gs://$GCS_BUCKET/dataset/mmlu/* ./artifacts/dataset/
+  fi
+
+  echo "Copying dataset to container..."
   docker cp artifacts/dataset "$CONTAINER_NAME:/workspace/"
```
Comment (QiliangCui) [inline scripts/agent/docker_run_bm.sh:105]: just to double confirm:  `docker cp` here can copy the whole folder and subfolder.

**Review by QiliangCui — APPROVED** [review]
